### PR TITLE
Extract the Hapi Server to expose it for testing

### PR DESCRIPTION
**Pull request by benstepp**
* The Hapi Server has been extracted so that it can be
tested using the built in #inject() method.
* Test support has been moved to /helpers. This is the default location
that is ignored by `ava`.
* Added a test to ensure that the user can override the default html.js
in both the develop and build stages.

After some research I think using Hapi's inject is going to provide the best testing experience with the least amount of test setup. This way we have simpler test infrastructure that is less likely to be the cause of test failures.

The only code change was that the server is now started in `lib/utils/develop` and constructed in `lib/utils/dev-server`

### CODE REVIEW

**Comment by benstepp** [comment]
I'll mess with babel tomorrow, I just wanted some feedback from CI. Maybe the require.resolve trick in the webpack config will take care of it.

**Comment by benstepp** [comment]
@KyleAMathews  I want to get some feedback on this approach. Basically we would be able to remove all of the babel stuff from starters because they don't need anything beyond what gatsby provides by default.

Rather than relying on babel's plugin resolving, we just pass resolved plugin paths to the babel-loader. A user could still override them with the modifyWebpackConfig API. 

* Easier on testing (no need to inject the webpack config into the hapi server)
* Easier on users (no need to add a babelrc or babel-plugins to package.json)
* Still allows for advanced babel config by overriding the loader

**Comment by KyleAMathews** [comment]
@benstepp potentially I really like this. I really don't like that users have to supply the Babel config/plugins right now. The only question I have here is how does this work (as in #235) when Gatsby is in a sub-directory of a Babel-driven JS module. Say someone building a React component wants to use Gatsby for their documentation site so inside Gatsby they're requiring up a directory into the src of the React component e.g. (`require('../../src/MyComponent')`). And say they're using a custom `.babelrc`. Do you know if Babel would still use their Babel setup with this change?

**Comment by benstepp** [comment]
tldr; babel takes anything for a plugin. we are just telling it explicitly rather than letting babel figure it out for us.

---

So the custom .babelrc is something I would like to support before this gets merged, as that would cause this to be a nonbreaking change. Basically this uses `require.resolve` to convert the babel-plugins into absolute paths relative to the gatsby installation. I ran into this idea when looking at a few issues to have the gatsby develop server use the gatsby's root level babel-plugins:

* https://github.com/babel/babel-loader/issues/149
* https://github.com/babel/babel-loader/issues/166

I went and looked at what babel allows for a plugin, and you can use pretty much anything (makes sense if maybe your company has private plugins for something). And it resolves them here:

* https://github.com/babel/babel/blob/master/packages/babel-core/src/transformation/file/options/option-manager.js#L145
* https://github.com/babel/babel/blob/master/packages/babel-core/src/helpers/resolve.js

The idea for supporting a custom .babelrc would be to:

1. check if the user has one
1. read the babelrc and JSON.parse it
1. map all of the plugins into absolute paths (so if you require 5 directories up the babel stuff still works because babel will treat them as absolute-path plugins and not auto resolve them in the directory that was crawled up to). As long as the .babelrc in the root of the gatsby project can cover whatever components are above, everything will work fine.
1. pass to webpack config.

If the user has a .babelrc they would have to also have any plugins they want to use specified locally (either in the gatsby directory or crawled up towards root. If we use `require.resolve` it will find them.) The user could copy their main top level .babelrc into the docs directory if they wanted.

We would have to change how react-hmre gets added as well to extend the user's babelrc rather than overwrite it. (probably with a `plugins.push(require.resolve('babel-plugin-react-hmre'))`) such that the user's babelrc is still relevant.

I have a strong feeling that this will also fix #129 by allowing the user to use whatever babel plugins they want. I'll explore some more tomorrow.

**Comment by benstepp** [comment]
Closing in favor of #279. There was a lot of noise here and I would rather do that separately from the babel changes.
